### PR TITLE
reflector: allow watch method to accept a watcher

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
This is a no-op PR. It paves the way for https://github.com/kubernetes/kubernetes/pull/110772.

It allows the watch method to accept a watcher.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 1]
```diff
@@ -366,35 +365,35 @@ func (r *Reflector) watch(stopCh <-chan struct{}, resyncerrc chan error) error {
...
+				// To reduce load on kube-apiserver on watch restarts, you may enable watch bookmarks.
+				// Reflector doesn't assume bookmarks are returned at all (if the server do not support
+				// watch bookmarks, it will ignore this field).
+				AllowWatchBookmarks: true,
+			}
+
+			w, err = r.listerWatcher.Watch(options)
+			if err != nil {
+				if canRetry := isWatchErrorRetriable(err); canRetry {
+					klog.V(2).Infof("%s: watch of %v returned %v - backing off", r.name, r.typeDescription, err)
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go:387]: This has a potential to be spammy if apiserver is overloaded.
Please make it V(4).

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 2]
```diff
@@ -366,35 +365,35 @@ func (r *Reflector) watch(stopCh <-chan struct{}, resyncerrc chan error) error {
...
+					<-r.initConnBackoffManager.Backoff().C()
+					continue
+				}
+				return err
 			}
-			return err
 		}
 
 		err = watchHandler(start, w, r.store, r.expectedType, r.expectedGVK, r.name, r.typeDescription, r.setLastSyncResourceVersion, r.clock, resyncerrc, stopCh)
+		w = nil // always create a new watch request
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go:396]: nit: let's maybe be a bit more explicit.
```
// Ensure that watch will not be reused across iterations.
w = nil
```

**Review by wojtek-t — COMMENTED** [review]
just two minor nits.

**Comment by p0lyn0mial** [comment]
@wojtek-t pushed, PTAL - thanks!

**Pull request merged**